sc-kpm/sc-agents-common/utils/AgentUtils.cpp

```diff
@@ -20,85 +21,104 @@ namespace utils
-    }
-    else break;
+    ScAddr edge = ms_context->CreateEdge(ScType::EdgeAccessConstPosPerm, questionNode, params.at(i));
+    ScAddr relation = IteratorUtils::getRoleRelation(ms_context, i + 1);
+    ms_context->CreateEdge(ScType::EdgeAccessConstPosPerm, relation, edge);
   }
 }
 
 ScAddr AgentUtils::createQuestionNode(ScMemoryContext * ms_context)
 {
   ScAddr questionNode = ms_context->CreateNode(ScType::NodeConst);
-  ScAddr question = CoreKeynodes::question;
-  ms_context->CreateEdge(ScType::EdgeAccessConstPosPerm, question, questionNode);
+  ms_context->CreateEdge(ScType::EdgeAccessConstPosPerm,  CoreKeynodes::question, questionNode);
   return questionNode;
 }
 
-bool AgentUtils::waitAgentResult(ScMemoryContext * ms_context, const ScAddr & questionNode)
+bool AgentUtils::waitAgentResult(ScMemoryContext * ms_context, const ScAddr & questionNode, const uint32_t waitTime)
 {
+  SC_CHECK_PARAM(questionNode, ("Invalid question node address"))
+
   auto check = [ms_context](ScAddr const & listenAddr, ScAddr const & edgeAddr, ScAddr const & otherAddr)
   {
     return ms_context->HelperCheckEdge(CoreKeynodes::nrel_answer, edgeAddr, ScType::EdgeAccessConstPosPerm);
   };
   ScWaitCondition<ScEventAddOutputEdge> waiter(*ms_context, questionNode, SC_WAIT_CHECK(check));
-  return waiter.Wait(30000);
+  return waiter.Wait(waitTime);
 }
 
-ScAddr AgentUtils::initAgent(ScMemoryContext * ms_context, const ScAddr & questionName, const vector<ScAddr> & params)
+ScAddr AgentUtils::initAgent(ScMemoryContext * ms_context, const ScAddr & questionClass, const ScAddrVector & params)
 {
+  SC_CHECK_PARAM(questionClass, ("Invalid question class address"))
+
   ScAddr questionNode = createQuestionNode(ms_context);
   assignParamsToQuestionNode(ms_context, questionNode, params);
-  ms_context->CreateEdge(ScType::EdgeAccessConstPosPerm, CoreKeynodes::question, questionNode);
+
+  ms_context->CreateEdge(ScType::EdgeAccessConstPosPerm, questionClass, questionNode);
   ms_context->CreateEdge(ScType::EdgeAccessConstPosPerm, CoreKeynodes::question_initiated, questionNode);
-  ms_context->CreateEdge(ScType::EdgeAccessConstPosPerm, questionName, questionNode);
   return questionNode;
 }
 
 ScAddr AgentUtils::initAgentAndWaitResult(
       ScMemoryContext * ms_context,
-      const ScAddr & questionName,
-      const vector<ScAddr> & params)
+      const ScAddr & questionClass,
+      const ScAddrVector & params)
 {
-  ScAddr questionNode = initAgent(ms_context, questionName, params);
-  ScAddr answer;
+  SC_CHECK_PARAM(questionClass, ("Invalid question class address"))
+
+  ScAddr questionNode = initAgent(ms_context, questionClass, params);
   if (waitAgentResult(ms_context, questionNode))
   {
     ScAddr nrelAnswer = CoreKeynodes::nrel_answer;
-    answer = IteratorUtils::getFirstByOutRelation(ms_context, questionNode, nrelAnswer);
+    return IteratorUtils::getAnyByOutRelation(ms_context, questionNode, nrelAnswer);
```

Comment:
Why do you prefer to use two returns? Let's discuss about this

Reply:
A dynamic control translation of gcc compiler is usually optimized with minimal variables in current stack